async-queues/src/lib.rs

```diff
@@ -0,0 +1,249 @@
+        unsafe { Box::from_raw(slice::from_raw_parts_mut(self.0.inner.data as *mut Slot<T>, self.0.inner.capacity as usize)) };
+    }
+}
+
+pub fn channel<T>(capacity: usize) -> (Sender<T>, Reciever<T>) {
+    let queue = Arc::new(Queue::new(capacity));
+    (Sender(queue.clone()), Reciever(queue))
+}
+
+/*
+/// A circular buffer used as a FIFO queue with atomic reads and writes.
+///
+/// The read offset is the element that was most recently read by the
+/// receiving end of the queue. The write offset is the element that was
+/// most recently written by the sending end. If the two offsets are equal,
+/// the queue is either empty or full.
+///
+/// The size of the buffer is such that not all the bits of the offset are
+/// necessary to encode the current offset. The next highest unused bit is
+/// used to keep track of the number of times the offset has wrapped
+/// around. If the offsets are the same and the bit is the same in the read
+/// and write offsets, the queue is empty. If the bit is different in the
+/// read and write offsets, the queue is full.
+///
+/// The following procedures will operate the queues in a multiple producer
+/// single consumer (MPSC) fashion.
+*/
+
+pub struct FifoDescriptorContainer<T> {
+    pub inner: FifoDescriptor<T>,
+}
+
+unsafe impl <T:Send> Sync for FifoDescriptorContainer<T> {}
+unsafe impl <T:Send> Send for FifoDescriptorContainer<T> {}
+
+#[derive(Debug)]
+struct Offsets {
+    read: u32,
+    write: u32,
+    read_blocked: bool,
+}
+
+impl Offsets {
+    fn as_usize(&self) -> usize {
+        ((self.write as usize) << 32) | (if self.read_blocked { 1 << 31 } else { 0 }) | (self.read as usize)
+    }
+
+    fn from_usize(offsets: usize) -> Self {
+        Offsets {
+            write: (offsets >> 32) as u32,
+            read_blocked: offsets & (1 << 31) != 0,
+            read: (offsets & ((1 << 31) - 1)) as u32,
+        }
+    }
+
+    fn is_empty(&self) -> bool {
+        self.read == self.write
+    }
+
+    fn is_full(&self, capacity: u32) -> bool {
+        self.read | capacity == self.write | capacity && !self.is_empty()
+    }
+
+    fn read_offset(&self, capacity: u32) -> isize {
+        (self.read & (capacity - 1)) as isize
+    }
+
+    fn write_offset(&self, capacity: u32) -> isize {
+        (self.write & (capacity - 1)) as isize
+    }
+
+    fn increment_read_offset(&mut self, capacity: u32) {
+        self.read = (self.read + 1) & ((capacity << 1) - 1);
+    }
+
+    fn increment_write_offset(&mut self, capacity: u32) {
+        self.write = (self.write + 1) & ((capacity << 1) - 1);
+        self.read_blocked = false;
+    }
+
+    fn len(&self, capacity: u32) -> usize {
+        ((if self.read <= self.write { 0 } else { 2 * capacity }) + self.write - self.read) as usize
+    }
+}
+
+impl<T: Copy> FifoDescriptorContainer<T> {
+    pub fn new(capacity: usize) -> Self {
+        assert!(capacity <= usize::pow(2, 31));
```

Comment:
In the documentation below you specify that the capacity must be a power of 2 less than `2^31` but here and I think in one other place you check only that it is a number less than `2^31`.